kimchi/src/circuits/polynomials/rot.rs

```diff
@@ -0,0 +1,608 @@
+
+        // Sum 2-bit limbs
+        for i in (7..COLUMNS).rev() {
+            bound += power_of_2.clone() * env.witness_curr(i);
+            power_of_2 *= T::from(4u64); // 2 bits
+        }
+
+        // Sum 12-bit limbs
+        for i in (3..=6).rev() {
+            bound += power_of_2.clone() * env.witness_curr(i);
+            power_of_2 *= 4096u64.into(); // 12 bits
+        }
+
+        // Check that excess < 2^rot by checking that bound < 2^64
+        // C10:bound = excess - 2^rot + 2^64
+        constraints.push(bound - (excess - two_to_rot + two_to_64));
+
+        constraints
+    }
+}
+
+// ROTATION WITNESS COMPUTATION
+
+fn layout_rot64<F: PrimeField>(curr_row: usize) -> [[Box<dyn WitnessCell<F>>; COLUMNS]; 2] {
+    [rot_row(), range_check_0_row("shifted", curr_row + 1)]
+}
+
+fn rot_row<F: PrimeField>() -> [Box<dyn WitnessCell<F>>; COLUMNS] {
+    [
+        VariableCell::create("word"),
+        VariableCell::create("rotated"),
+        VariableCell::create("excess"),
+        /* 12-bit plookups */
+        VariableBitsCell::create("bound", 52, Some(64)),
+        VariableBitsCell::create("bound", 40, Some(52)),
+        VariableBitsCell::create("bound", 28, Some(40)),
+        VariableBitsCell::create("bound", 16, Some(28)),
+        /* 2-bit crumbs */
+        VariableBitsCell::create("bound", 14, Some(16)),
+        VariableBitsCell::create("bound", 12, Some(14)),
+        VariableBitsCell::create("bound", 10, Some(12)),
+        VariableBitsCell::create("bound", 8, Some(10)),
+        VariableBitsCell::create("bound", 6, Some(8)),
+        VariableBitsCell::create("bound", 4, Some(6)),
+        VariableBitsCell::create("bound", 2, Some(4)),
+        VariableBitsCell::create("bound", 0, Some(2)),
+    ]
+}
+
+fn init_rot64<F: PrimeField>(
+    witness: &mut [Vec<F>; COLUMNS],
+    curr_row: usize,
+    word: F,
+    rotated: F,
+    excess: F,
+    shifted: F,
+    bound: F,
+) {
+    let rot_rows = layout_rot64(curr_row);
+    witness::init(
+        witness,
+        curr_row,
+        &rot_rows,
+        &variable_map!["word" => word, "rotated" => rotated, "excess" => excess, "shifted" => shifted, "bound" => excess+bound],
+    );
+}
+
+/// Extends the rot rows to the full witness
+pub fn extend_rot_rows<F: PrimeField>(
+    witness: &mut [Vec<F>; COLUMNS],
+    word: F,
+    rotated: F,
+    excess: F,
+    shifted: F,
+    bound: F,
+) {
+    let rot_row = witness[0].len();
+    let rot_witness: [Vec<F>; COLUMNS] = array::from_fn(|_| vec![F::zero(); 2]);
+    for col in 0..COLUMNS {
+        witness[col].extend(rot_witness[col].iter());
+    }
+    init_rot64(witness, rot_row, word, rotated, excess, shifted, bound);
+}
+
+/// Create a rotation witness
+/// Input: word to be rotated, rotation offset, and side of rotation.
+/// Output: witness for rotation word and initial row with all zeros
+pub fn create_witness<F: PrimeField>(word: u64, rot: u32, side: RotMode) -> [Vec<F>; COLUMNS] {
```

Comment:
Do you allow `rot = 0` or  `rot > 64` for this pub interface?  If not, maybe check it or assert like you did for the other interface.